docs/ccv1_image_security_design.md

```diff
@@ -0,0 +1,403 @@
+# CCv1 Image Security Design
+
+This document describes the design of container image security under the confidential container V1 architecture, including the encryption-decryption and signing of container image. The components involved include image-rs, ocicrypt-rs, attestation-agent and Key Broker Service.
+
+## Production of protected container image
+
+In the confidential containers V1 design, after the user completes the production of the conventional container image, it is necessary to add security protection functions for the container image, including encryption and signing. It should be noted that the encryption process must be carried out before the signing process.
+
+The format standard of container image encryption is compatible with https://github.com/containers/ocicrypt.
+
+The format standard of container image signing is compatible with https://github.com/containers/image.
+
+### Image encryption
+
+Image encryption is based on layer granularity. Users can selectively encrypt layers containing private data or choose not to encrypt any layers. The high-level process of encrypting an image layer is as follows:
```

Comment:
Didn't we conclude at some point late last year that our security model might not be robust for images with no encryption at all? @fitzthum were you the one raising that point (sorry, my memory is a bit fuzzy)?

Reply:
I'm also unclear about images containing both encrypted and unencrypted layers. How can CC guarantee safety in the following sort of scenario?:

| layer | encrypted? | dependent layer |
|-|-|-|
| 1 | yes | n/a |
| 2 | no | n/a |
| 3 | yes | 2 |

Reply:
cc @jiazhang0

Reply:
@jodh-intel To me, the guarantee that you have in that case is that you won't be able to execute anything in that image before you decrypted the encrypted layers, which requires data from the attestation itself. With only signed layers, at least conceptually, you could run the image without going through attestation.

Reply:
Because we don't explicitly verify the identity of the KBS, signature verification is only guaranteed by the successful decryption of an encrypted layer. In other words, your workload is secure only if its execution is gated by the secret. Note that this is scoped at the pod level, so you can trust signed but unencrypted containers as long as something in the pod depends on depends on the secret. This should be a pretty good fit to workloads that use some containers that access secret data and some generic containers or one container with some generic layers and some secret layers. 

Also note that we do signature verification at the container level, so I don't think we can have a mix of signed and unsigned layers within one container.

Reply:
@c3d @jodh-intel 

Hopefully this explanation can help you to understand why image encryption is useful: https://github.com/confidential-containers/documentation/issues/18 《What image encryption means for CC？》. Briefly, image encryption can enforce to inject an explicit authorization during CC bootstrap through Attestation Evidence for Authorization, causing a Pod has to request Relying Party to verify its TCB state.

Reply:
> I'm also unclear about images containing both encrypted and unencrypted layers. How can CC guarantee safety in the following sort of scenario?:
> layer 	encrypted? 	signed? 	dependent layer
> 1 	yes 	yes 	n/a
> 2 	no 	no 	n/a
> 3 	yes 	yes 	2

First, image signing takes its effect on image manifest digest (in other word, the entire image) rather than an individual layer. You could refer to https://github.com/confidential-containers/documentation/issues/18 《What container image signing means for CC？》for more explanations about it.

Reply:
@jiazhang0 Thanks for the pointer to the documentation. My question was not about the usefulness of encryption (I have no doubt about it :smile:) but about the weakness of the attestation and workload process if we have only signature.

As I wrote in my comment, the risk is running a workload without a valid attestation when you don't need any secret to decrypt the image. @fitzthum reminded me of one easy method to achieve that, which is to insert a bogus KBS in the loop.

As he also pointed out, image decryption is not the only step that may require a secret. If the workload has some encrypted volume being used only by super-standard containers (e.g. some `mysql`), then arguably signature is sufficient, since what you really want to protect is the data, and that data is not accessible without a secret.

In short, I'm trying to evaluate if supporting a non-encrypted signed image makes sense or not. At the moment, I believe there are valid use cases, as long as it does not break the rest

Reply:
> As I wrote in my comment, the risk is running a workload without a valid attestation when you don't need any secret to decrypt the image. @fitzthum reminded me of one easy method to achieve that, which is to insert a bogus KBS in the loop.

To accurately answer this, we need to distinguish the use cases according to boot image strategy. For protected boot image strategy, the TCB state of Pod is trusted. It is safe to verify a signed but unencrypted image (w/o sensitive data) because all artifacts used to verify the signature of image and the verification result are all trusted. For unprotected boot image strategy, the initial TCB state of Pod is unverified, so the initial image to be pulled must be an encrypted, which will enforce an remote attestation to let remote party verify the TCB state of Pod and then unseal the decryption key. Once the TCB state of Pod was verified, the following images to be pulled can be unencrypted as long as it doesn't contain sensitive data.

So the TCB state of Pod is important to conclude whether unencrypted image has weakness or not. Essentially speaking, whether the verification result of signed image is trusted is determined by the TCB state of Pod. If it is unverified or untrusted, the verification result or even whether the signed image is actually occurred is doubtful. The introduction of image encryption can introduce an opportunity to check the TCB stage of Pod.

> In short, I'm trying to evaluate if supporting a non-encrypted signed image makes sense or not. At the moment, I believe there are valid use cases, as long as it does not break the rest

Yes. For sidecar containers, it doesn't contain sensitive data, so it is not necessary to be encrypted.

Reply:
We did talk in the one of calls about the possibility in the future of having another component which might have wider awareness of protected v unprotected boot strategy, TCB state of the POD, status of all container images so far with respect to signatures and encryption and where signatures/keys originated from  (i.e. are they all unencrypted so far or has one been decrypted). 
This component can combine all this knowledge to take a view if the Pod has any unaddressed risk that might impact security/confidentiality.

Bottom line is that something somewhere needs to depend upon a key that is either secure in hardware (closer to protected bootimage) or provided by KBS as part of the attestation flow (could be for container decryption but could also be key for data decryption - application won't work without the data). If all these are untrue then we have a problem, if one is true then answer could be "depends"

Reply:
> We did talk in the one of calls about the possibility in the future of having another component which might have wider awareness of protected v unprotected boot strategy, TCB state of the POD, status of all container images so far with respect to signatures and encryption and where signatures/keys originated from (i.e. are they all unencrypted so far or has one been decrypted). This component can combine all this knowledge to take a view if the Pod has any unaddressed risk that might impact security/confidentiality.
> 

I guess that component is Root of Chain for Measurement (RTM), consisting of the initial measurement, dynamic measurement or vTPM (w/o dynamic measurement), and event log.